### ISSUE TITLE
[PRE REVIEW]: haldensify: Highly adaptive lasso conditional density estimation in R

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2389]
**Submitting author:** @nhejazi (<a href="http://orcid.org/0000-0002-7127-2789">Nima Hejazi</a>)
**Repository:** <a href="https://github.com/nhejazi/haldensify" target ="_blank">https://github.com/nhejazi/haldensify</a>
**Version:** v0.0.6
**Editor:** Pending
**Reviewer:** Pending
**Managing EiC:** Arfon Smith

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**Author instructions**

Thanks for submitting your paper to JOSS @nhejazi. **Currently, there isn't an JOSS editor assigned** to your paper.

@nhejazi if you have any suggestions for potential reviewers then please mention them here in this thread (without tagging them with an @). In addition, [this list of people](https://bit.ly/joss-reviewers) have already agreed to review for JOSS and may be suitable for this submission (please start at the bottom of the list).

**Editor instructions**

The JOSS submission bot @whedon is here to help you find and assign reviewers and start the main review. To find out what @whedon can do for you type:

```
@whedon commands
```

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.84  T=0.11 s (416.4 files/s, 63703.5 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
HTML                            16            708            184           3071
R                                8             75            329            451
CSS                              3             99             48            428
Markdown                         4            106              0            275
JavaScript                       3             64             32            256
TeX                              1             12              0            128
YAML                             5             29              2            104
Rmd                              2             75            259             81
XML                              1              0              0             27
make                             1              9              0             20
SVG                              1              0              1             11
-------------------------------------------------------------------------------
SUM:                            45           1177            855           4852
-------------------------------------------------------------------------------


Statistical information for the repository '2389' was gathered on 2020/06/25.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Nima Hejazi                      3           364             12          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Nima Hejazi                 352           96.7          0.1                9.09
```

**Comment by whedon** [comment]
PDF failed to compile for issue #2389 with the following error: 

 Can't find any papers to compile :-(

**Comment by arfon** [comment]
@whedon generate pdf from branch joss_paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss_paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02389/joss.02389/10.21105.joss.02389.pdf)

**Comment by arfon** [comment]
:wave: @nhejazi - Thanks for your submission to JOSS. After consulting with the JOSS editorial team, we've determined that this work is out of scope for JOSS as it falls into our [minor utility category](https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements). 

One possible alternative to JOSS is to follow [GitHub's guide](https://guides.github.com/activities/citable-code/) on how to create a permanent archive and DOI for your software. This DOI can then be used by others to cite your work.

**Comment by arfon** [comment]
@whedon reject

**Comment by whedon** [comment]
Paper rejected.

**Comment by nhejazi** [comment]
Hi @arfon, and JOSS editorial team, thanks for your quick initial review of our `haldensify` R package. If possible, would it be possible to clarify exactly how the determination was made that this package falls into the minor utility category?

As it's not at this point clear what exactly this initial review process included, I clarify below why the package should not fall into the minor utility designation as well as the scope of the package:

- `haldensify` is neither a thin API around any other single package/program nor a single-function utility package. Although these are but examples, minor utility packages would generally be expected to provide "convenience" functionality, that is, they ought to produce results (in this case, numerical) that would be otherwise easy to generate with only a bit of extra work, e.g., perhaps a nicer grammar for querying a database or a trivial transformation of the results produced by another package.
- `haldensify` exposes six functions (https://code.nimahejazi.org/haldensify/reference/index.html), several of which may be of independent interest (e.g., `fit_haldensify`, `cv_haldensify`) for various applications in conditional density estimation.
- Given that it may have been determined that `haldensify` is but a thin wrapper around the `hal9001` package, it's worth clarifying that this is not the case. Where `hal9001` provides functionality for fitting a highly adaptive lasso regression model, `haldensify` implements a novel use of this regression model in producing conditional density estimates through discretization. In our experience, this was not trivial -- in fact, it includes (1) restructuring the input data through a unique discretization (https://code.nimahejazi.org/haldensify/reference/format_long_hazards.html); (2) external cross-validation (that implemented in `hal9001` is incompatible) to select tuning parameters based on a loss function appropriate for density estimation; (3) re-scaling estimates from the hazard scale to the density scale.
- In addition to the above, `haldensify` implements an option for fitting a sequence of conditional density models that may be used in novel selection procedures to construct statistically efficient estimators. To the best of our knowledge, this is a previously unexplored area of statistical research in the context of conditional density estimators, and one on which we are actively finishing a statistical methods manuscript.
- Finally, to the best of our knowledge, only two previous attempts have been made to implement any element of the functionality in `haldensify`: (1) the original paper (Diaz & van der Laan, 2011) did not produce a software package; (2) another package linked to in our README tried to implement this methodology, but included a critical bug (which actually led to our need to write `haldensify`), and, even then, did not implement the general framework in the context of the highly adaptive lasso estimator.

**Comment by arfon** [comment]
Hi @nhejazi, thanks for your response here, this additional context is very useful.

Your submission was flagged primarily because of its small size (~450 lines of R). While lines of code is a rather coarse measure for evaluating scale/impact/complexity of a software project it is _something_, and part of our role as an editorial team is to maintain some kind of 'level of effort' threshold for submissions that we publish.

I recognize that there's not really anything in our submission guidelines about this, and the JOSS editorial team are currently writing up our thoughts on this complex topic (blog post coming soon at http://blog.joss.theoj.org).

I note you have a second submission (https://github.com/openjournals/joss-reviews/issues/2383) on a related package. Might it make sense to write a single paper describing both of these projects together? Having multiple software projects reviewed together in a single JOSS submission is something we've handled before.

**Comment by nhejazi** [comment]
Thanks for the quick reply, @arfon; I'm glad we can continue this discussion. Thanks also for explaining the reason `haldensify` was flagged as possibly out of scope. I'm sure there are at least a few checks that go into ensuring that a given software project is of reasonable effort to be reviewed and have a corresponding paper published. My own view here would be that lines of code as a metric alone is a _very_ coarse metric; even in addition to commit history, lines of code over that commit history, etc., there's a lot that would be missing in putting together a picture of the scope of a package.

By such metrics, it's easy to miss packages that follow the philosophy that a package should do a single thing well -- that is, for example, within `haldensify`, there would have been a lot of opportunity to write boilerplate code (e.g., creating custom classes, adding many accessor functions to interact with slots of these classes, etc.) that would have significantly increased the lines of code, possibly satisfying that metric...but many of these additions would have little improved the package. All of this is to say that, of course, more lines of code is not necessarily more effort / more functionality. That said, I realize there does need to be some metric, and, aside from things like number of commits, how long a project has been around, whether it's part of a larger repository that enforces background quality checks (for R, CRAN or Bioconductor), I don't have much to add in terms of ideas.

I didn't check how the ~450 lines of R code was calculated, but, at a glance, I'm seeing ~700 lines in the R subdirectory + another ~350 lines between the unit tests and vignettes (maybe the check ignores documentation/examples/tests? but I would argue it shouldn't as these are certainly not low-effort). As mentioned above, `haldensify` implements an adaptation of a particular conditional density estimation approach -- one that is not widely known and not particularly easy to implement (i.e., there wasn't a working implementation in the 10-11 years between when the methods paper was published and when `haldensify` was first built; otherwise, I wouldn't have written this package). On that account alone, I'd think it would be an acceptable contribution.

Finally, thanks for the comment about the separate submission of the `hal9001` package. Unfortunately, these packages are too different in scope for `haldensify` to be merged into that package, I think. `hal9001` started out a few years ago as a collaboration that implemented the HAL regression model and that has since made several efficiency improvements to that approach. `haldensify` is an effort that's about a year old and part of a different collaboration; as such, it differs significantly, both in context and scope, from the `hal9001` package. While `haldensify` relies upon `hal9001`, it uses that regression algorithm only for a single step in it's much more involved process to produce conditional density estimates. I'd also expect the corresponding software papers for each package to be cited in distinct projects, so it might be a bit odd to have a single paper on these two very different algorithms. For what it's worth, these packages being submitted at approximately the same time to JOSS is moreso an artifact of my only having time to get around to the submissions now (vs. a year or a few months ago).

**Comment by arfon** [comment]
Hi @nhejazi, apologies for the slow response here. 

I wanted to share a couple of extra resources that weren't public until yesterday that articulate some of our thinking on these topics about scope/level of effort in JOSS submissions:

- https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort
- https://blog.joss.theoj.org/2020/07/minimum-publishable-unit

As you can see from these, we're trying to ensure that submissions to JOSS meet a certain threshold of effort, and believe that this is absolutely essential to maintaining the long-term reputation of the journal (and therefore delivering on our mission to provide a mechanism for career credit for researchers).

_More thoughts specifically on your submission(s)_

> As mentioned above, `haldensify` implements an adaptation of a particular conditional density estimation approach -- one that is not widely known and not particularly easy to implement (i.e., there wasn't a working implementation in the 10-11 years between when the methods paper was published and when `haldensify` was first built; otherwise, I wouldn't have written this package). On that account alone, I'd think it would be an acceptable contribution.

I have no doubt that this is a very useful contribution. My concern here is whether it meets the [substantial scholarly effort test](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort). As you are probably aware, [_Salami slicing_](https://en.wikipedia.org/wiki/Least_publishable_unit) of research outputs is an issue that communities (and journals) need to be on the look out for, and while I'm *not* accusing you of this here, two related submissions in quick succession did necessitate a review on our part.

> I note you have a second submission (#2383) on a related package. Might it make sense to write a single paper describing both of these projects together? Having multiple software projects reviewed together in a single JOSS submission is something we've handled before.

☝️ my comment here was really not about asking you to merge the two software packages together, rather, I was trying to explore whether a single JOSS paper could be use to describe both packages together?

Also, I note that in your paper you write _"[...] may be estimated by an arbitrary machine learning algorithm"._  Myself and a few other JOSS editors were wondering if it would be an idea to make a more general package "...densify" that could use different methods, e.g. `hal9001`?

**Comment by nhejazi** [comment]
No problem at all, I appreciate the continued discussion @arfon. Thank you for pointing out the blog post and documentation describing JOSS's updated submission/review policy. The explanation provided in these write-ups is helpful, and I'm glad to see that JOSS editorial team is taking steps to ensure that publications in the journal can continue to be relied upon for their quality.

Based on the blog post, it seems that there might still be an avenue for having `haldensify` reviewed by JOSS. My reasoning for this is that `haldensify` contains ~450 lines of R code and ~80 lines of Rmd, not including code comments, documentation, or prose explaining the functionality of the package. Based on our discussion, I think the package has room for a few improvements that may add another ~100 lines of R code and some more Rmd documentation explaining new features. While the package is unlikely to reach 1000+ lines of R code (at that point, it'll probably be close to ~700-750 lines of R code), I do think a few more features might make the contribution more significant.

Continuing our discussion

> I have no doubt that this is a very useful contribution. My concern here is whether it meets the substantial scholarly effort test. As you are probably aware, Salami slicing of research outputs is an issue that communities (and journals) need to be on the look out for, and while I'm not accusing you of this here, two related submissions in quick succession did necessitate a review on our part.

I agree that the two submissions in quick succession should have triggered some manual review mechanism. Unfortunately, submission of the `hal9001` and `haldensify` papers in this short window was a consequence of the availability of my time: both had been on my todo list for some time and I was able to carve out just enough time to get to both only recently. From looking at the commit histories, these packages are quite different
  
- `hal9001` is a larger effort, started by myself and Jeremy Coyle (@jeremyrcoyle) in August 2017. Since reaching maturity, it's been relied upon in several publications, though it has not always been cited. A few recent instances where I've cited it myself include (1) https://arxiv.org/abs/1901.02776, (2) https://arxiv.org/abs/1912.09936, and (3) https://arxiv.org/abs/2005.11303. The package is currently relied upon as a core component of our research program and used across many projects, including `haldensify`.
- `haldensify` is a much smaller package, only ever intended to support conditional density estimation with the highly adaptive lasso. It started as an offshoot of a project that David Benkeser (@benkeser) and I were working on since ~2017, though it only became necessary to implement this methodology ourselves towards the end of 2018 and into early 2019, when we realized that the existing implementation in https://github.com/osofr/condensier did not work (https://github.com/osofr/condensier/issues/15). Since then, it's grown to support using another type of highly adaptive lasso procedure for constructing efficient estimators when conditional density estimates are a nuisance parameter. While it relies upon `hal9001`, it generally implements a different algorithm, first written about in https://www.degruyter.com/view/journals/ijb/7/1/article-0000102202155746791356.xml. We needed this procedure for the work we did in https://arxiv.org/abs/2003.13771, and we are currently completing a methodological paper based largely on `haldensify`. I want to emphasize that this is _not salami slicing_, as these are two entirely separate techniques. _Essentially, we found ourselves in the unfortunate position of realizing that a tool we intended to use didn't work and weren't able to fix said tool; so, we decided to implement what we did need, which did not exist in any form at the time._

> :point_up: my comment here was really not about asking you to merge the two software packages together, rather, I was trying to explore whether a single JOSS paper could be use to describe both packages together?

Hopefully my descriptions of the separate origins of these two projects clarifies why I don't think this would be a suitable solution. The projects are separate, implement different algorithms, and are intended for different purposes. They have different goals, different histories, and may diverge even further in the future.

> Also, I note that in your paper you write "[...] may be estimated by an arbitrary machine learning algorithm". Myself and a few other JOSS editors were wondering if it would be an idea to make a more general package "...densify" that could use different methods, e.g. `hal9001`?

Thanks for raising this point. Indeed, the general algorithm implemented in `haldensify` can be used outside of HAL regression via `hal9001`. From statistical theory, HAL regression has a few special properties that make it a good choice for estimating nuisance parameters of the estimators that our group typically works with; the same properties don't necessarily hold for other nonparametric regression procedures like random forests or xgboost. Thus, we decided to restrict this package just to HAL regression. I wouldn't want to generalize the package to allow arbitrary estimation since the same theoretical guarantees cannot be made for other procedures (I should have explained this better in the paper). This question did give me an idea (referenced above) for improving the package though. To make the package a better contribution, I can add an inverse probability weighted estimator (originally from https://onlinelibrary.wiley.com/doi/full/10.1111/j.1541-0420.2011.01685.x) that requires conditional density estimates from `haldensify`; this would be a novel statistical advance, as it would leverage theory from https://arxiv.org/abs/2005.11303 to provide an estimator that is improved relative to what https://onlinelibrary.wiley.com/doi/full/10.1111/j.1541-0420.2011.01685.x originally formulated. The `haldensify` package would then provide two useful procedures that are outside of the scope of `hal9001`. I'd be happy for any comments on these ideas to enhance the package. If the editorial team feels that `haldensify` would not be suitable for JOSS even with these changes, please do let me know.
